File: client/dart/lib/src/transaction.dart
import 'package:trailbase/src/client.dart';

class Operation {
  _CreateOperation? create;
  _UpdateOperation? update;
  _DeleteOperation? delete;

  Operation({this.create, this.update, this.delete});

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    if (create != null) {
      data['Create'] = create!.toJson();
    }
    if (update != null) {
      data['Update'] = update!.toJson();
    }
    if (delete != null) {
      data['Delete'] = delete!.toJson();
    }
    return data;
  }
}

class _CreateOperation {
  String apiName;
  Map<String, dynamic> value;

  _CreateOperation({required this.apiName, required this.value});

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['api_name'] = apiName;
    data['value'] = value;
    return data;
  }
}

class _UpdateOperation {
  String apiName;
  RecordId recordId;
  Map<String, dynamic> value;

  _UpdateOperation({
    required this.apiName,
    required this.recordId,
    required this.value,
  });

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['api_name'] = apiName;
    data['record_id'] = recordId.toString();
    data['value'] = value;
    return data;
  }
}

class _DeleteOperation {
  String apiName;
  RecordId recordId;

  _DeleteOperation({required this.apiName, required this.recordId});

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['api_name'] = apiName;
    data['record_id'] = recordId.toString();
    return data;
  }
}

class TransactionRequest {
  List<Operation> operations;

  TransactionRequest({required this.operations});

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = {};
    data['operations'] = operations.map((e) => e.toJson()).toList();
    return data;
  }
}

abstract class ITransactionBatch {
  IApiBatch api(String apiName);
  Future<List<RecordId>> send();
}

abstract class IApiBatch {
  ITransactionBatch create(Map<String, dynamic> value);
  ITransactionBatch update(RecordId recordId, Map<String, dynamic> value);
  ITransactionBatch delete(RecordId recordId);
}

class TransactionBatch implements ITransactionBatch {
  static const String _transactionApi = 'api/transaction/v1/execute';

  final Client _client;
  final List<Operation> _operations = [];

  TransactionBatch(this._client);

  @override
  IApiBatch api(String apiName) {
    return ApiBatch(this, apiName);
  }

  @override
  Future<List<RecordId>> send() async {
    final request = TransactionRequest(operations: _operations);
    final response = await _client.fetch(
      TransactionBatch._transactionApi,
      method: 'POST',
      data: request.toJson(),
    );

    if ((response.statusCode ?? 400) > 200) {
      throw Exception('${response.data} ${response.statusMessage}');
    }

    final result = ResponseRecordIds.fromJson(response.data);
    return result.toRecordIds();
  }

  void _addOperation(Operation operation) {
    _operations.add(operation);
  }
}

class ApiBatch implements IApiBatch {
  final TransactionBatch _batch;
  final String _apiName;

  ApiBatch(this._batch, this._apiName);

  @override
  ITransactionBatch create(Map<String, dynamic> value) {
    _batch._addOperation(
      Operation(create: _CreateOperation(apiName: _apiName, value: value)),
    );
    return _batch;
  }

  @override
  ITransactionBatch update(RecordId recordId, Map<String, dynamic> value) {
    _batch._addOperation(
      Operation(
        update: _UpdateOperation(
          apiName: _apiName,
          recordId: recordId,
          value: value,
        ),
      ),
    );
    return _batch;
  }

  @override
  ITransactionBatch delete(RecordId recordId) {
    _batch._addOperation(
      Operation(
          delete: _DeleteOperation(apiName: _apiName, recordId: recordId)),
    );
    return _batch;
  }
}

File: client/swift/trailbase/Sources/TrailBase/Transaction.swift
import Foundation

public enum Operation: Codable {
    case create(apiName: String, value: [String: AnyCodable])
    case update(apiName: String, recordId: RecordId, value: [String: AnyCodable])
    case delete(apiName: String, recordId: RecordId)

    private enum CodingKeys: String, CodingKey {
        case create = "Create"
        case update = "Update"
        case delete = "Delete"
        case apiName = "api_name"
        case recordId = "record_id"
        case value = "value"
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)

        switch self {
        case let .create(apiName, value):
            var createContainer = container.nestedContainer(keyedBy: CodingKeys.self, forKey: .create)
            try createContainer.encode(apiName, forKey: .apiName)
            try createContainer.encode(value, forKey: .value)

        case let .update(apiName, recordId, value):
            var updateContainer = container.nestedContainer(keyedBy: CodingKeys.self, forKey: .update)
            try updateContainer.encode(apiName, forKey: .apiName)
            try updateContainer.encode("\(recordId)", forKey: .recordId)
            try updateContainer.encode(value, forKey: .value)

        case let .delete(apiName, recordId):
            var deleteContainer = container.nestedContainer(keyedBy: CodingKeys.self, forKey: .delete)
            try deleteContainer.encode(apiName, forKey: .apiName)
            try deleteContainer.encode("\(recordId)", forKey: .recordId)
        }
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)

        if let createContainer = try? container.nestedContainer(keyedBy: CodingKeys.self, forKey: .create) {
            let apiName = try createContainer.decode(String.self, forKey: .apiName)
            let value = try createContainer.decode([String: AnyCodable].self, forKey: .value)
            self = .create(apiName: apiName, value: value)

        } else if let updateContainer = try? container.nestedContainer(keyedBy: CodingKeys.self, forKey: .update) {
            let apiName = try updateContainer.decode(String.self, forKey: .apiName)
            let recordIdString = try updateContainer.decode(String.self, forKey: .recordId)
            let value = try updateContainer.decode([String: AnyCodable].self, forKey: .value)
            self = .update(apiName: apiName, recordId: RecordId.string(recordIdString), value: value)

        } else if let deleteContainer = try? container.nestedContainer(keyedBy: CodingKeys.self, forKey: .delete) {
            let apiName = try deleteContainer.decode(String.self, forKey: .apiName)
            let recordIdString = try deleteContainer.decode(String.self, forKey: .recordId)
            self = .delete(apiName: apiName, recordId: RecordId.string(recordIdString))

        } else {
            throw DecodingError.dataCorrupted(
                DecodingError.Context(codingPath: container.codingPath, debugDescription: "Invalid Operation type")
            )
        }
    }
}

public struct TransactionRequest: Codable {
    public var operations: [Operation] = []

    private enum CodingKeys: String, CodingKey {
        case operations = "operations"
    }
}

public struct TransactionResponse: Codable {
    public var ids: [String] = []

    private enum CodingKeys: String, CodingKey {
        case ids = "ids"
    }
}

public class TransactionBatch {
    private let client: Client
    private var operations: [Operation] = []

    init(client: Client) {
        self.client = client
    }

    public func api(_ apiName: String) -> ApiBatch {
        return ApiBatch(batch: self, apiName: apiName)
    }

    public func send() async throws -> [RecordId] {
        let request = TransactionRequest(operations: operations)
        let body = try JSONEncoder().encode(request)

        let (_, data) = try await client.fetch(
            path: "api/transaction/v1/execute",
            method: "POST",
            body: body
        )

        let response = try JSONDecoder().decode(TransactionResponse.self, from: data)
        return response.ids.map { RecordId.string($0) }
    }

    internal func addOperation(_ operation: Operation) {
        operations.append(operation)
    }
}

public class ApiBatch {
    private let batch: TransactionBatch
    private let apiName: String

    init(batch: TransactionBatch, apiName: String) {
        self.batch = batch
        self.apiName = apiName
    }

    public func create(value: [String: AnyCodable]) -> TransactionBatch {
        batch.addOperation(.create(apiName: apiName, value: value))
        return batch
    }

    public func update(recordId: RecordId, value: [String: AnyCodable]) -> TransactionBatch {
        batch.addOperation(.update(apiName: apiName, recordId: recordId, value: value))
        return batch
    }

    public func delete(recordId: RecordId) -> TransactionBatch {
        batch.addOperation(.delete(apiName: apiName, recordId: recordId))
        return batch
    }
}

public struct AnyCodable: Codable {
    public let value: Any

    public init<T>(_ value: T?) {
        self.value = value ?? ()
    }

    public init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self.value = ()
        } else if let bool = try? container.decode(Bool.self) {
            self.value = bool
        } else if let int = try? container.decode(Int.self) {
            self.value = int
        } else if let double = try? container.decode(Double.self) {
            self.value = double
        } else if let string = try? container.decode(String.self) {
            self.value = string
        } else if let array = try? container.decode([AnyCodable].self) {
            self.value = array.map { $0.value }
        } else if let dictionary = try? container.decode([String: AnyCodable].self) {
            self.value = dictionary.mapValues { $0.value }
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "AnyCodable value cannot be decoded")
        }
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        if value is () {
            try container.encodeNil()
        } else if let bool = value as? Bool {
            try container.encode(bool)
        } else if let int = value as? Int {
            try container.encode(int)
        } else if let double = value as? Double {
            try container.encode(double)
        } else if let string = value as? String {
            try container.encode(string)
        } else if let array = value as? [Any] {
            try container.encode(array.map(AnyCodable.init))
        } else if let dictionary = value as? [String: Any] {
            try container.encode(dictionary.mapValues(AnyCodable.init))
        } else {
            let context = EncodingError.Context(codingPath: container.codingPath, debugDescription: "AnyCodable value cannot be encoded")
            throw EncodingError.invalidValue(value, context)
        }
    }
}
